includes/admin/load.php

```diff
@@ -204,51 +210,75 @@ function perflab_dismiss_wp_pointer_wrapper() {
-	);
+/**
+ * Register admin actions for handling performance feature plugin installation/activation.
+ *
+ * @since n.e.x.t
+ */
+function perflab_register_admin_actions() {
+	add_action( 'admin_action_perflab_install_activate_plugins', 'perflab_install_activate_plugins_callback' );
 }
+add_action( 'admin_init', 'perflab_register_admin_actions' );
 
 /**
- * Callback function hooked to admin_action_perflab_activate_plugin to handle plugin activation.
+ * Callback for handling installation/activation of plugin.
  *
- * @since 2.8.0
+ * @since n.e.x.t
  */
-function perflab_activate_plugin() {
-	// Do not proceed if plugin query arg is not present.
-	if ( empty( $_GET['plugin'] ) ) {
-		return;
+function perflab_install_activate_plugins_callback() {
+	check_admin_referer( 'perflab_install_activate_plugins' );
+
+	require_once ABSPATH . 'wp-admin/includes/plugin-install.php';
+	require_once ABSPATH . 'wp-admin/includes/class-wp-upgrader.php';
+	require_once ABSPATH . 'wp-admin/includes/class-wp-ajax-upgrader-skin.php';
+
+	if ( ! isset( $_GET['slug'] ) ) {
+		wp_die( esc_html__( 'Missing required parameter.', 'performance-lab' ) );
+	}
+
+	if ( ! current_user_can( 'install_plugins' ) || ! current_user_can( 'activate_plugins' ) ) {
+		wp_die( esc_html__( 'Sorry, you are not allowed to manage plugins for this site. Please contact the administrator.', 'performance-lab' ) );
+	}
+
+	$plugin = sanitize_text_field( wp_unslash( $_GET['slug'] ) );
+
+	if ( ! $plugin ) {
+		wp_die( esc_html__( 'Invalid plugin.', 'performance-lab' ) );
 	}
 
-	$plugin = sanitize_text_field( wp_unslash( $_GET['plugin'] ) );
+	$api = perflab_query_plugin_info( $plugin );
 
-	check_admin_referer( "perflab_activate_plugin_{$plugin}" );
+	// Return early if plugin API return an error.
+	if ( ! $api ) {
+		wp_die( esc_html__( 'An unexpected error occurred. Something may be wrong with WordPress.org or this server&#8217;s configuration.', 'performance-lab' ) );
+	}
+
+	$plugins             = get_plugins();
+	$is_plugin_installed = ! ( isset( $plugins[ $plugin . '/load.php' ] ) || isset( $plugins[ $plugin . '/' . $plugin . '.php' ] ) );
+	$status              = array();
+
+	// Install the plugin if it is not installed yet.
+	if ( $is_plugin_installed ) {
+		// Replace new Plugin_Installer_Skin with new Quiet_Upgrader_Skin when output needs to be suppressed.
+		$skin     = new WP_Ajax_Upgrader_Skin( array( 'api' => $api ) );
+		$upgrader = new Plugin_Upgrader( $skin );
+		$result   = $upgrader->install( $api['download_link'] );
+
+		if ( is_wp_error( $result ) ) {
+			wp_die( esc_html( $result->get_error_message() ) );
+		} elseif ( is_wp_error( $skin->result ) ) {
+			wp_die( esc_html( $skin->result->get_error_message() ) );
+		} elseif ( $skin->get_errors()->has_errors() ) {
+			wp_die( esc_html( $skin->get_error_messages() ) );
+		}
+	}
 
 	// If `$plugin` is a plugin slug rather than a plugin basename, determine the full plugin basename.
 	if ( ! str_contains( $plugin, '/' ) ) {
```

Comment:
This check isn't useful now because `$plugin` is _always_ the slug based on your logic changes. Instead, related to my other feedback, we should check whether the passed `file` is empty, and if so use the `slug` to determine the file / basename below.